### PR TITLE
Openrefine

### PR DESCRIPTION
**Pull request by wolfv**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

File: recipes/openrefine/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+build:
+  number: 0
+  skip: true  # [not linux]
+
+requirements:
+  run:
+    - openjdk
+
+test:
+  commands:
```
Comment (scopatz) [inline recipes/openrefine/meta.yaml:23]: This recipe needs some sort of test

**Comment by scopatz** [comment]
This seems to be a binary repack, is that rigth @wolfv?

**Comment by scopatz** [comment]
Is there some issue with building it?

**Comment by wolfv** [comment]
yep it's a repack -- i have never built a java package so I don't know how complicated that would be?

**Comment by scopatz** [comment]
I think if it is pure Java that is OK.  But this seems like it might contain compiled code. CC @mariusvniekerk?

**Comment by scopatz** [comment]
Also, I think you just have to use maven to build in most cases

**Comment by mariusvniekerk** [comment]
It's got a pom.xml so you can just build it with maven

**Comment by wetneb** [comment]
I would recommend to use `mvn compile test-compile` instead of `mvn install`. Installing OpenRefine in Maven's repository will not do anything interesting and might be the cause of the error you have.

For me, this release still compiles fine:
```
git checkout 3.3
mvn clean
mvn compile test-compile
```
(or `./refine build` works too).

File: recipes/openrefine/meta.yaml (hunk 1 above)
Comment (wetneb) [reply recipes/openrefine/meta.yaml:23]: `./refine test` or `mvn test` should do

**Comment by wolfv** [comment]
thanks, I'll try that!

btw. do you happen to know the right mvn instruction to install the final files into a Prefix (e.g. $PREFIX/bin/openrefine etc.).

**Comment by wetneb** [comment]
> btw. do you happen to know the right mvn instruction to install the final files into a Prefix (e.g. $PREFIX/bin/openrefine etc.).

We don't have any at the moment I am afraid.

**Comment by wolfv** [comment]
I think you still have the packages in your cache, which is why it works. The URL for butterfly does return a 404 for me here: 

https://repo.maven.apache.org/maven2/edu/mit/simile/butterfly/1.0.2/butterfly-1.0.2.pom

**Comment by wetneb** [comment]
ah right you also need to install the packages we ship locally. `./refine build` does it for you, but if you want to use Maven directly you can do so with `mvn process-resources` first.

**Comment by wolfv** [comment]
@wetneb thanks, let's see :) are you interested in being a co-maintainer of this package? That usually means that once the feedstock is created, a bot makes PRs to update the versions and sometimes minor things need to be fixed.

**Comment by wetneb** [comment]
I'd be happy to be listed yes!

**Comment by wolfv** [comment]
cool the build goes through (on Linux, will enable the other platforms now). Will add you as a maintainer.

Do you know which files we need to copy into the $PREFIX?

**Comment by wetneb** [comment]
The files required to launch OpenRefine are described in the Linux packaging profile, in `packaging/linux.xml`

**Comment by wolfv** [comment]
locally I ran `mvn install -DskipTests` and that worked (created a tarball for Linux). Now I just need to move the contents of that tar ball to the right place I believe.

File: recipes/openrefine/build.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# mkdir -p $PREFIX/bin/
+# mv refine* $PREFIX/opt/openrefine/
+# mv server $PREFIX/opt/openrefine/server
+# mv webapp $PREFIX/opt/openrefine/webapp
+
+# cat > $PREFIX/bin/refine <<EOL
+# $PREFIX/opt/openrefine/refine "$@"
+# EOL
+
+# chmod u+x $PREFIX/bin/refine
```
Comment (scopatz) [inline recipes/openrefine/build.sh:18]: Please remove dead code here

**Comment by wetneb** [comment]
(by the way we released 3.4 a few days ago so it could be worth updating it here)

**Comment by wolfv** [comment]
thanks for the pointer @wetneb 

btw is that true that mvn install creates a tarball, adn I basically have to unzip the tarball into the right place afterwards?

**Comment by wetneb** [comment]
`mvn package` creates a zip archive, yes, and the user can unzip that anywhere and run OpenRefine from there.

**Comment by wolfv** [comment]
I see that it creates a zip for win, a tar.gz for linux and a dmg file for OS X.

I can handle the tar.gz and the zip.. and for OS X I will use hditool to mount the dmg file and copy the contents from there? Does that make sense?

**Comment by wetneb** [comment]
That makes perfect sense I would say!

**Comment by wolfv** [comment]
@wetneb are the contents of the tar.gz the same vs. dmg? might be easier to extract the tar.gz on both OS X and Linux if yes.

**Comment by wetneb** [comment]
@wolfv correct, the linux bundle can be run on Mac too, as long as a JRE is present (we bundle the JRE in the DMG package).

**Comment by wolfv** [comment]
it looks like the `mvn package` step on Windows doesn't produce the `.zip` file... Maybe for the time being we should repack the non-JRE release for Windows, then...

**Comment by wolfv** [comment]
I feel the whole idea of building this package from source is somewhat fruitless. Ok if I repackage the built releases from GitHub? @scopatz @mariusvniekerk ?

**Comment by mariusvniekerk** [comment]
If you can get just the Java bits without the jre that sounds fine to me.

**Comment by wetneb** [comment]
It is curious that Windows `.zip` archives are not produced when building on Windows.

If you are happy with binaries then yes we can use the official `.zip` for windows (without JRE) and the `.tar.gz` for Linux and Mac (without JRE either).

**Comment by wolfv** [comment]
I just tried again with windows, locally, and it does indeed seem to build the zip file (and the other distribution packages) correctly when I execute it locally. Sorry for the false alarm.
I am still going to go for the repackaging right now, because otherwise we'd need to switch to a newer docker image for Linux on Azure as far as I understand.

**Comment by wolfv** [comment]
Ok from my perspective this recipe is ready now!
I have left dead code / comments in so that we can re-enable the maven based build on which I spent considerable time once this glibc issue is resolved.

**Comment by scopatz** [comment]
I think this should wait until the glibc issues have been resolved then. I don't think we should be doing a binary repack just out of expediency.

**Comment by wolfv** [comment]
there is no native code nor a glibc dependency in the resulting JAR files as far as I understand. Just some transient build dependency, when downloaded through maven, seems to have a symbol issue. So I don't think your comment applies here.

**Comment by wolfv** [comment]
Also just FYI openrefine is an _awesome_ web app that would be a great addition to the conda-forge data science ecosystem.

**Comment by wolfv** [comment]
@scopatz the glibc issue is in a vendored `windres` binary https://sourceware.org/binutils/docs/binutils/windres.html

This is probably used to build the Windows package on Linux (so not necessary in our case), since we can build linux on linux, windows on windows etc.

I could try to figure out how maven works, and try to disable the build of the windows package on Linux. 
Or we can merge, and move to centos7 once the feedstock is created.

I am still convinced that this is good to go as the resulting zip files from mvn package should be identical to what we can get from github. I am happy to iterate in the feedstock.

**Comment by wolfv** [comment]
The `windres` binary is vendored by launch4j

**Comment by scopatz** [comment]
Yeah, we probably need a few more people who know about the Java ecosystem.

**Comment by wetneb** [comment]
It is useful to know that launch4j is causing some troubles on your side, we will definitely take that into account for our packaging improvements.

In the meantime, I do not know if it would be acceptable to just change the packaging configuration to disable the Windows build in the Linux environment? It is not exactly pretty, but as far as I can tell Maven does not have any CLI option for that sadly.

Here is a patch to do so:
[0001-Remove-Windows-packaging-profile.patch.tar.gz](https://github.com/conda-forge/staged-recipes/files/5219891/0001-Remove-Windows-packaging-profile.patch.tar.gz)

**Comment by wolfv** [comment]
@wetneb thanks, I was trying to do that, but I don't know anything about maven ...

**Comment by wolfv** [comment]
now i need to figure out how to get `genisoimage` on our linux docker image to build the dmg for OS X I guess (or can we disable that on Linux, too?)

**Comment by wolfv** [comment]
I think I am getting a little closer: there are assembly files that we can run, and 

`mvn assembly:assembly ...` does almost the right thing. I am trying to figure out how to only run the assembly step so that it puts the files that end up in the `.zip` or `.tar.gz` file uncompressed into some directory.

When I try to run `mvn assembly:assembly` it tells me

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.2-beta-5:assembly (default-cli) on project openrefine: Error reading assemblies: No assembly descriptors found. -> [Help 1]
```

Maybe we can set some option for this to take the files from packaging?

**Comment by wolfv** [comment]
there are also: assembly, directory-single, attached, single, help, unpack, directory-inline

**Comment by wetneb** [comment]
Since you are using the Linux build for both Mac and OSX I would indeed disable the call to genisoimage indeed. You could do it similarly to the patch to remove the Windows build. It should be enough to just remove the `<plugin></plugin>` tag that starts with `<groupId>sh.tak.appbundler</groupId>`.

**Comment by wolfv** [comment]
turns out the reason why no `.zip` file was created on windows was that `mvn` is a Batch file itself, and running the first `mvn` command terminates the batch script (it works when using `CALL mvn ...`.

Now I think win should be fine as well.

File: recipes/openrefine/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+source:
+  url: https://github.com/OpenRefine/OpenRefine/archive/3.4.tar.gz
+  sha256: ffea300cb5d4fc0b78423ac78b71f954c6addb8490bdd2ee66e58bab993a2154
+  patches:
+    - 0001-Remove-Windows-packaging-profile.patch  # [linux]
+  # url: https://github.com/OpenRefine/OpenRefine/releases/download/{{ version }}/openrefine-linux-{{ version }}.tar.gz  # [unix]
+  # sha256: fe7b40604b325d31b60042ce3df73b2af2cbe48d0c4687cc1a4934d760560d66  # [unix]
+  # url: https://github.com/OpenRefine/OpenRefine/releases/download/{{ version }}/openrefine-win-{{ version }}.zip  # [win]
+  # sha256: ee592124a3bde2ab9ed511f7eb5c938ea7aa41503a73d79d877fb67aa3432405  # [win]
+  # folder: openrefine
```
Comment (scopatz) [inline recipes/openrefine/meta.yaml:17]: Please remove old sources

File: recipes/openrefine/bld.bat [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+CALL mvn process-resources
+CALL mvn package -DskipTests
+
+CD %SRC_DIR%\packaging\target\
+tar -xvf openrefine-win-%PKG_VERSION%.zip
+
+ROBOCOPY openrefine-%PKG_VERSION% %LIBRARY_PREFIX%\opt\openrefine /E
+
+ECHO CD %LIBRARY_PREFIX%\opt\openrefine\ > %LIBRARY_PREFIX%\bin\refine.bat
+ECHO CALL refine.bat %%* >> %LIBRARY_PREFIX%\bin\refine.bat
```
Comment (scopatz) [inline recipes/openrefine/bld.bat:14]: Please add exit conditions

**Comment by wolfv** [comment]
@scopatz this should be ready to go now :)

**Pull request merged**